### PR TITLE
1T Likelihood

### PR DESCRIPTION
**Pull request by maspalla**
## Description

We added a custom Likelihood function reconstructing a single hadronic top. 

## Motivation and Context

Needed for the KLFitter implementation we use in our ttbar+bbar ATLAS analysis.

## How Has This Been Tested?

Extensively used in our analysis, the code was inside the ATLAS software package.
Once the function has been copied to the KLFitter branch, the code compiles.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by knutzk** [comment]
Hi @maspalla,

Thanks a lot for this PR! It's great to see people are still using KLFitter and that it sees active development. I would like to trigger a discussion on a couple points before moving forward with this PR:
1. Just so the new likelihood follows our general coding style, I have made a few changes that I would like you to add to this PR. For that, I'm gonna pull request into the branch of your fork – once you accept that, this PR updates automatically. Would that be ok with you?
2. Once (1) is done, we can then discuss any remaining issues with the code in-line.
3. In terms of physics: am I correct in understanding that this likelihood reconstructs _one single top quark_ with no additional  particles? What exactly is the physics motivation behind that? From your description I'm assuming you're targeting ttbar – is your assumption then that the other top quark is not recoverable?
4. How well is this likelihood validated? If you have any public material, it would be very helpful to see this here: reconstruction efficiencies, computing times, validation plots, etc.

Thanks a lot!

File: CMakeLists.txt [hunk 1]
```diff
@@ -32,7 +32,7 @@ include( CTest )
 # Set up the "C++ version" to use.
 set( CMAKE_CXX_STANDARD_REQUIRED 11 CACHE STRING
    "Minimum C++ standard required for the build" )
-set( CMAKE_CXX_STANDARD 14 CACHE STRING
+set( CMAKE_CXX_STANDARD 17 CACHE STRING
```
Comment (knutzk) [inline CMakeLists.txt:35]: I'm assuming this was by mistake? If there are good reasons to change the CXX standard, we can certainly discuss about it, but it should not be part of this PR then.

**Review by knutzk — CHANGES_REQUESTED** [review]

**Comment by maspalla** [comment]
Hi @knutzk,

thanks for the detailed checks, I'll comment on specific comments separately.

Concerning the physics motivation, this likelihood indeed reconstructs only 1 top. 
The reason behind it is to try to speed up the all-hadronic ttbar reconstruction by computing the likelihood for 1 top at a time and later building the ttbar likelihood by simply summing them up.
This was needed because the KLFitter implementation we developed in our analysis relies on multiple calls of the KLFitter and was quite time consuming. This 'one top' approach reduces a lot the number of calls to the fitter and make the run much more time efficient. 

Unfortunately I cannot explain here our KLFitter implementation, because non of it is public, and would also make for quite a complex discussion, but let’s make a more general example.
Imagine we have an event with 10 jets, 2 b-tags and we are running KLFitter in the all-hadronic mode and with bVeto. One wants to loop over all possible jet combinations, compute the likelihood and find the best one.
(Our method is actually more complicated than this, but this gives an idea of the reasoning)
Since the fitter is called once per permutation, one can compute the expected number of fitter calls in such a scenario.
This would correspond to 2 x b(8,4) x b(4,2) = 840 2T KLFitter calls for the standard all-hadronic approach, but only 2 x b(8,2) = 56 with this single-top likelihood.
2 x is the choice of b with the selected top in both cases. b(n,m) is the binomial coefficient choosing m out of n when the order does not matter and b(8,4) is the number of unique choices of 4 light-flavour in 8 light-flavour jets, while b(4,2) is the number of unique 2 W you can form from 4 light-flavour jets. b(8,2) is the number of single W to form from 8 light-flavour jets.

This likelihood has been tested extensively, in order to check if we get the same results as with regular all-hadronic. Again, I cannot put the plots here, but we are confident it can be used without any problems.

However, I personally believe this could be used also to target cases in which only one top can be reconstructed, even though it is not our aim.

Best,
Margherita

File: CMakeLists.txt (hunk 1 above)
Comment (maspalla) [reply CMakeLists.txt:35]: Hi @knutzk ,
this was actually needed to be able to compile within our athena setup. It failed with CXX 14.
However, we of course don't aim at compiling this ourselves within athena, but rather have it at central level and use that one. I don't really know what works there.
You're welcome to have it back to what is ok for standalone KLFitter.
Best,
M.

**Review by knutzk — APPROVED** [review]

File: include/KLFitter/LikelihoodSingleTopAllHadronic.h [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+  /* @{ */
+
+  /**
+   * Set the values for the missing ET x and y components and the m_et_miss_sum. This
+   * sets the internal variables #m_et_miss_x, #m_et_miss_y and #m_et_miss_sum to the
+   * given values.
+   * @param etx missing ET x component.
+   * @param ety missing ET y component.
+   * @param sumet total scalar ET.
+   * @return An error flag.
```
Comment (knutzk) [inline include/KLFitter/LikelihoodSingleTopAllHadronic.h:128]:
```suggestion
   * For events with missing ET this function would set the x and y components. In this
   * class, the function is only implemented as a trivial function as MET is not used
   * in the likelihood.
```
Sorry, I think I copy-pasted this comment blindly from the l+jets likelihood. The description for this particular likelihood would be rather confusing, so I would suggest the above change.

**Review by knutzk — CHANGES_REQUESTED** [review]
Hi @maspalla,

Thanks for accepting my proposed code changes. I made one suggested change above about one of the comments that I accidentally added to the code. I also checked whether the code documentation gets picked up correctly for http://klfitter.github.io/ and that seems to work fine, too.

Otherwise, I think the only thing that's left would be to add a brief documentation of this likelihood to the README files of KLFitter in a similar fashion as for the other likelihoods: https://github.com/KLFitter/KLFitter/blob/master/doc/WhatIsKLF.md#reconstruction-of-top-quark-events Maybe you could just add a section below the "Single-top t+W events" subsection and a paragraph or two about this particular likelihood too? That would be great!

Once that is done, I think we can merge this PR and create a new KLFitter version tag.

**Comment by maspalla** [comment]
Hi @knutzk ,
ok, I'll add some description in the readme.
Best,
M.

**Comment by knutzk** [comment]
Hi @maspalla,

That would be great, thanks! Last but not least, I discussed with @TomasDado about the naming of this likelihood, and we think that both "SingleTop" as well as "AllHadronic" might not be entirely appropriate – the first because it suggests a relation to single-top final states (for which "SgTop" had already been used for a different likelihood), the latter because the final state is just "hadronic". Could we converge on something like "LikelihoodHadronicTop"? 

Best,
Knut

**Comment by maspalla** [comment]
Hi @knutzk ,

I agree about the name, but can I suggest "LikelihoodHadronic1Top" or "LikelihoodHadronicOneTop" ?
It emphasizes that it's only 1 top, without mentioning the word 'single'.
Besides, we call it '1T' in our documentation, so it would be all consistent.
What do you think?

Best,
M.

**Comment by knutzk** [comment]
LikelihoodHadronicOneTop sounds good to me – let's go with that!

**Comment by maspalla** [comment]
Hi @knutzk ,
I just pushed the changes we discussed but I'm going to have to change a couple more things in the documentation text tomorrow.
Can you please wait a bit before finalising the pull request?
Thanks.
Best,
M.

**Comment by maspalla** [comment]
Hi @knutzk ,
ok, ready to merge from our side.

Best,
M.

**Comment by knutzk** [comment]
Hi @maspalla,

Thanks for the updates! The README addition reads well to me, but I noticed that you re-committed the _old_ version of the likelihood under its name, i.e. without my suggested changes. If it's fine for you, I will merge by going back to original proposal + my edits + name change. Please let me know!

Knut

**Review by knutzk — APPROVED** [review]

**Comment by maspalla** [comment]
Hi @knutzk ,
of course the idea was to have the name change on top of your suggestions, didn't mean to change back to the old likelihood.
Just for my understanding, I'm a bit confused: you mean that LikelihoodOneHadronicTop has the old namespace usage, variable names etc.?
Thanks.
M.

**Comment by knutzk** [comment]
Yes, exactly! All changes from maspalla/KLFitter#1 are gone in the renamed files. I can fix that easily and prepare the merge, I just need your o.k.

**Comment by maspalla** [comment]
Hi @knutzk ,
ah ok! Sure, please go ahead!
Sorry for the mess.

Best,
M.

**Comment by knutzk** [comment]
Hi @maspalla, this got merged in #101. Thanks for the contribution! Can you also please let us know who we should list in [doc/Authors.md] for preparing this likelihood (including affiliation)? Thanks a lot!

We'll get back to you as soon as we have prepared a new version release – which can then get propagated to the ATLAS software infrastructure.

Knut

**Comment by knutzk** [comment]
Hi @maspalla, just a quick reminder to please let us know who to put on the contributor list for the likelihood. I'm waiting for that to prepare the new KLFitter tag.

**Comment by maspalla** [comment]
Hi @knutzk ,

sorry for the long wait.

Since this was very much a team effort, we were thinking of something along the line of ‘Margherita Spalla on behalf of the ttbb analysis team’.
Would that work with your policies for KLFitter?

Thanks.

Best,
M.

**Comment by knutzk** [comment]
Hi @maspalla,

thanks! I don’t see any issues in terms of policy, but be aware that no one outside the collaboration will understand what it means, nor know who else might have been part of that team. If possible, I would be explicit and put individual names – but it’s up to you. Please let us know and I’ll update the readme files accordingly!

Knut

**Comment by maspalla** [comment]
Hi @knutzk ,

ok, I see. 
Then you can put 
T. Barillari, S. Menke, M. Spalla, N. Wenke
(order's alphabetic, as usual).

Best,
M.

**Comment by knutzk** [comment]
Hi @maspalla,

Sorry for being late myself this time! I added your proposed contribution in #102 and now prepared a new release tagged v1.4.0 (see https://github.com/KLFitter/KLFitter/releases). We're gonna go ahead with the version update within ATLAS now.

Cheers,
Knut

**Comment by maspalla** [comment]
Great, thanks a lot!
Best,
M.

**Pull request closed without merging**